linera-execution/src/system.rs

```diff
@@ -647,44 +648,96 @@ where
+                        self.balance.set(new_balance);
+                    } else {
+                        let balance = self.balances.get_mut_or_default(&target).await?;
+                        *balance = balance.saturating_add(amount);
+                    }
+                    Ok(None)
+                } else {
+                    // Handle cross-chain transfer with message.
+                    let message = SystemMessage::Credit {
+                        amount,
+                        source,
+                        target: account.owner,
+                    };
+                    Ok(Some(
+                        OutgoingMessage::new(account.chain_id, message)
+                            .with_kind(MessageKind::Tracked),
+                    ))
+                }
             }
             Recipient::Burn => Ok(None),
         }
     }
 
     pub async fn claim(
-        &self,
+        &mut self,
         authenticated_signer: Option<AccountOwner>,
         authenticated_application_id: Option<ApplicationId>,
         source: AccountOwner,
         target_id: ChainId,
         recipient: Recipient,
         amount: Amount,
-    ) -> Result<OutgoingMessage, ExecutionError> {
+    ) -> Result<Option<OutgoingMessage>, ExecutionError> {
         ensure!(
             authenticated_signer == Some(source)
                 || authenticated_application_id.map(AccountOwner::from) == Some(source),
             ExecutionError::UnauthenticatedClaimOwner
         );
         ensure!(amount > Amount::ZERO, ExecutionError::IncorrectClaimAmount);
 
-        let message = SystemMessage::Withdraw {
-            amount,
-            owner: source,
-            recipient,
-        };
-        Ok(
-            OutgoingMessage::new(target_id, message)
-                .with_authenticated_signer(authenticated_signer),
-        )
+        let current_chain_id = self.context().extra().chain_id();
+        if target_id == current_chain_id {
+            // Handle same-chain claim locally by processing the withdraw operation directly
+            self.debit(&source, amount).await?;
+            match recipient {
+                Recipient::Account(account) => {
+                    if account.chain_id == current_chain_id {
+                        // Handle same-chain recipient locally
+                        let target = account.owner;
+                        if target == AccountOwner::CHAIN {
+                            let new_balance = self.balance.get().saturating_add(amount);
+                            self.balance.set(new_balance);
+                        } else {
+                            let balance = self.balances.get_mut_or_default(&target).await?;
+                            *balance = balance.saturating_add(amount);
+                        }
+                    } else {
+                        // Recipient is on a different chain, create outgoing Credit message
+                        let message = SystemMessage::Credit {
```

Comment:
Either we shouldn't allow this, or we should merge the `transfer` and `claim` functions (and operations?) into one.